src/main/java/ru/ralsei/whatcanyousee/gameactivity/InternetConnector.java

```diff
@@ -0,0 +1,267 @@
+package ru.ralsei.whatcanyousee.gameactivity;
+
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import com.google.android.gms.games.multiplayer.Multiplayer;
+import com.google.android.gms.games.multiplayer.realtime.OnRealTimeMessageReceivedListener;
+import com.google.android.gms.games.multiplayer.realtime.RealTimeMessage;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+/**
+ * Class for managing internet messaging between player's.
+ */
+class InternetConnector {
+    private String TAG = "What can you see: Internet connector";
+
+    private GameActivity activity;
+
+    InternetConnector(GameActivity activity) {
+        this.activity = activity;
+    }
+
+    /**
+     * True if got message from teammate that he is ready to get a voice connection.
+     */
+    private boolean otherPlayerIsReady = false;
+
+    /**
+     * True if ready to send and receive voice.
+     */
+    private boolean prepared = false;
+
+    /**
+     * Reset prepared status for the next created game.
+     */
+    void clearResources() {
+        otherPlayerIsReady = false;
+        prepared = false;
+    }
+
+    /**
+     * Send to other player that we are ready to receive his voice.
+     * Also sends game settings if we are the game host.
+     */
+    void sendReadyMessage() {
+        byte[] message;
+
+        GameplayHandler.GameSettings gameSettings = activity.getGameplayHandler().getGameSettings();
+        if (gameSettings != null) {
+            try {
+                ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+                ObjectOutputStream writeStream;
+
+                writeStream = new ObjectOutputStream(byteStream);
+
+                gameSettings.flipSettings();
+                writeStream.writeObject(gameSettings);
+                writeStream.flush();
+                gameSettings.flipSettings();
+
+                message = byteStream.toByteArray();
+                writeStream.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+                activity.getUIHandler().showGameError();
+                activity.getGooglePlayHandler().leaveRoom();
+                return;
+            }
+        } else {
+            message = new byte[1];
+            message[0] = 'R';
+        }
+
+        if (message.length > Multiplayer.MAX_RELIABLE_MESSAGE_LEN) {
+            Log.d(TAG, "Ready message: message is too long.");
+        }
+
+        sendReliableMessage(message);
+    }
+
+    private void reactOnReceivedMessage(byte[] receivedData) {
+        if (!otherPlayerIsReady) {
```

Comment:
Inventing your own message protocol can be very error-prone when you need to extend it, when you have dynamic field lengths, etc. I suggest taking advantage of Java serialization and sending, for instance, serialized objects that look like this:

class Message {
  MessageType type;
  byte[] payload;
}

MessageType would be an enum denoting what has happened (other player won; you won; lever pressed etc.)
If the message has extra data, it can be serialized again and sent as bytes in the payload. Then, knowing the MessageType, the client will know what class to deserialize the payload into.

This schema is far from ideal, but at least it provides some type safety, which is completely lacking when you are using characters to denote different events.

Reply:
Done, but there is a little problem: Java serialization provides a lot of trash info (i.e. class descriptions), so using it for all messaging seems to me as a really bad idea considering overwhelming majority of sending data is the voice audio (same MessageType).
So I made a little compromise: voice audio data starts with 'P', while serialized objects guaranteed (by Java implementation) to start with bytes AC ED.